File: cpp/143_Reorder_List.cpp
/*
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You may not modify the values in the list's nodes, only nodes itself may be changed.

Example 1:
Given 1->2->3->4, reorder it to 1->4->2->3.

Example 2:
Given 1->2->3->4->5, reorder it to 1->5->2->4->3.
*/
#include <iostream>

using namespace std;

class Solution {
    public:
        struct ListNode {
            int val;
            ListNode *next;
            ListNode(int x) : val(x), next(NULL) {}
        };

        void reorderList(ListNode* head) {
            if(!head || !head->next || !head->next->next) return;
            ListNode *p1 = head, *p2 = head;
            while(p2->next && p2->next->next) {
                p1 = p1->next;
                p2 = p2->next->next;
            }
            // cut off from middle
            ListNode *head2 = p1->next;
            p1->next = NULL;
            ListNode *temp, *pre = NULL;    // *pre = NULL ensures the correctness of list length
            while(head2) {
                temp = head2->next;
                head2->next = pre;
                pre = head2;
                head2 = temp;
            }
            p1 = head; p2 = pre;
            ListNode *t;
            while(p1) {
                t = p1->next;
                p1->next = p2;
                p1 = p1->next;
                p2 = t;
            }
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/143_Reorder_List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next or not head.next.next:
            return
        p1, p2 = head, head
        while p2.next and p2.next.next:
            p1 = p1.next
            p2 = p2.next.next
        p2 = p1.next
        p1.next = None

        prev, cur = None, p2
        while cur:
            tmp = cur.next
            cur.next = prev
            prev = cur
            cur = tmp

        p1, p2 = head, prev
        while p1:
            tmp = p1.next
            p1.next = p2
            p1 = p1.next
            p2 = tmp
        return head
